src/async_fuse/memfs/cache/memory_cache.rs

```diff
@@ -0,0 +1,649 @@
+    ///
+    /// A block may be evicted to the backend.
+    async fn write_block_into_cache(&self, ino: INum, block_id: usize, block: Block)
+    where
+        P: EvictPolicy<BlockCoordinate> + Send + Sync,
+        S: Storage + Send + Sync,
+    {
+        let mut retry_times = 0;
+
+        let mut file_cache = loop {
+            // TODO: returns error instead of `panic`.
+            assert!(
+                retry_times < Self::INSERT_RETRY_LIMMIT,
+                "Gave up retrying to insert a block into the cache."
+            );
+
+            let file_cache = {
+                let guard = pin();
+                Arc::clone(self.map.get_or_insert(ino, Arc::default(), &guard)).write_owned()
+            }
+            .await;
+
+            let success = self.policy.try_put(BlockCoordinate(ino, block_id));
+
+            if success {
+                break file_cache;
+            }
+            drop(file_cache);
+            self.evict().await;
+            retry_times = retry_times.overflow_add(1);
+        };
+
+        file_cache.insert(block_id, block);
+    }
+}
+
+#[async_trait]
+impl<P, S> Storage for InMemoryCache<P, S>
+where
+    P: EvictPolicy<BlockCoordinate> + Send + Sync,
+    S: Storage + Send + Sync,
+{
+    async fn load_from_self(&self, ino: INum, block_id: usize) -> Option<Block> {
+        let res = self.get_block_from_cache(ino, block_id).await;
+        if res.is_some() {
+            self.policy.touch(&BlockCoordinate(ino, block_id));
+        }
+        res
+    }
+
+    /// Loads a block from the backend.
+    ///
+    /// If `backend` returns a `None`, this will create a new block with zeros filled,
+    /// inserts it into the `backend`, and returns it to caller. That means, this method
+    /// will never return `None`.
+    async fn load_from_backend(&self, ino: INum, block_id: usize) -> Option<Block> {
+        if let Some(from_backend) = self.backend.load(ino, block_id).await {
+            Some(from_backend)
+        } else {
+            // If a fetching from backend still fails, create a new block filled with zeros,
+            // then write it to cache and backend as stored.
+            let mut zeroed_block = Block::new(self.block_size);
+
+            // This block is to be written
+            zeroed_block.set_dirty();
+
+            self.backend
+                .store(
+                    ino,
+                    block_id,
+                    IoBlock::new(zeroed_block.clone(), 0, self.block_size),
+                )
+                .await;
+            Some(zeroed_block)
+        }
+    }
+
+    async fn cache_block_from_backend(&self, ino: INum, block_id: usize, block: Block) {
+        self.write_block_into_cache(ino, block_id, block).await;
+    }
+
+    async fn store(&self, ino: INum, block_id: usize, mut input: IoBlock) {
+        let start_offset = input.offset();
+        let end_offset = input.end();
+
+        assert!(end_offset <= self.block_size, "out of range");
```

Comment:
use `debug_assert!` instead

Reply:
This is intentional but not for debugging, and will be replaced with an error returned in the future (the instant when error handling is introduced).

Reply:
Link to the issue #425